lib/imapurl.c

```diff
@@ -449,43 +448,43 @@ EXPORTED int imapurl_fromURL(struct imapurl *url, const char *s)
     return 0;
 }
 
-EXPORTED void imapurl_toURL(char *dst, const struct imapurl *url)
+EXPORTED void imapurl_toURL(struct buf *dst, const struct imapurl *url)
 {
 
     if (url->server) {
-        dst += sprintf(dst, "imap://");
-        if (url->user) dst += sprintf(dst, "%s", url->user);
-        if (url->auth) dst += sprintf(dst, ";AUTH=%s", url->auth);
+        buf_appendcstr(dst, "imap://");
+        if (url->user) buf_appendcstr(dst, url->user);
+        if (url->auth) buf_printf(dst, ";AUTH=%s", url->auth);
         if (url->user || url->auth)
-            *dst++ = '@';
-        dst += sprintf(dst, "%s", url->server);
+            buf_putc(dst, '@');
+        buf_appendcstr(dst, url->server);
     }
     if (url->mailbox) {
-        *dst++ = '/';
+        buf_putc(dst, '/');
         MailboxToURL(dst, url->mailbox);
-        dst += strlen(dst);
     }
 
     if (url->uidvalidity)
-        dst += sprintf(dst, ";UIDVALIDITY=%lu", url->uidvalidity);
+        buf_printf(dst, ";UIDVALIDITY=%lu", url->uidvalidity);
     if (url->uid) {
-        dst += sprintf(dst, "/;UID=%lu", url->uid);
-        if (url->section) dst += sprintf(dst, "/;SECTION=%s", url->section);
+        buf_printf(dst, "/;UID=%lu", url->uid);
+        if (url->section) buf_printf(dst, "/;SECTION=%s", url->section);
         if (url->start_octet || url->octet_count) {
-            dst += sprintf(dst, "/;PARTIAL=%lu", url->start_octet);
-            if (url->octet_count) dst += sprintf(dst, ".%lu", url->octet_count);
+            buf_printf(dst, "/;PARTIAL=%lu", url->start_octet);
+            if (url->octet_count) buf_printf(dst, ".%lu", url->octet_count);
         }
     }
     if (url->urlauth.access) {
         if (url->urlauth.expire) {
-            strcpy(dst, ";EXPIRE=");
-            dst += strlen(dst);
-            dst += time_to_iso8601(url->urlauth.expire, dst, INT_MAX, 1);
+            buf_appendcstr(dst, ";EXPIRE=");
+            char buf[RFC3339_DATETIME_MAX+1] = { 0 };
+            time_to_iso8601(url->urlauth.expire, buf, RFC3339_DATETIME_MAX, 1);
```

Comment:
Since time_to_iso8601() returns the length written, we *could* expand the dst buf by RFC3339_DATETIME_MAX, write directly into it at the current offset, and then truncate it to the correct length.

Reply:
I thought to keep this part of the code to be as simple as possible.